// @farcaster/snap@2.6.4/dist/react-native/components/snap-item-group.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Children, Fragment } from "react";
import { StyleSheet, View } from "react-native";
import { useSnapTheme } from "../theme.js";
import { SnapItemGroupBorderProvider } from "./item-layout-context.js";
const GAP_MAP = {
    none: 0,
    sm: 4,
    md: 8,
    lg: 12,
};
export function SnapItemGroup({ element: { props }, children, }) {
    const { colors } = useSnapTheme();
    const border = Boolean(props.border);
    const separator = Boolean(props.separator);
    const explicitGap = typeof props.gap === "string" ? String(props.gap) : undefined;
    const defaultGap = border || separator ? "sm" : "none";
    const gap = GAP_MAP[explicitGap ?? defaultGap] ?? GAP_MAP[defaultGap];
    const items = Children.toArray(children);
    return (_jsx(SnapItemGroupBorderProvider, { value: border, children: _jsx(View, { style: [
                styles.group,
                border && {
                    borderWidth: 1,
                    borderColor: colors.border,
                    borderRadius: 12,
                },
                { gap },
            ], children: items.map((child, i) => (_jsxs(Fragment, { children: [separator && i > 0 && (_jsx(View, { style: { height: 1, backgroundColor: colors.border + "80" } })), child] }, i))) }) }));
}
const styles = StyleSheet.create({
    group: {
        width: "100%",
        overflow: "hidden",
    },
});
