// @farcaster/snap@2.6.4/dist/react/components/item.js
"use client";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Item, ItemContent, ItemTitle, ItemDescription, ItemActions, ItemMedia, } from "@neynar/ui/item";
import { cn } from "@neynar/ui/utils";
import { useSnapColors } from "../hooks/use-snap-colors.js";
import { useSnapStackDirection } from "../stack-direction-context.js";
import { ICON_MAP } from "./icon.js";
import { useSnapItemGroupHasBorder } from "./item-layout-context.js";
function parseItemMedia(value) {
    if (!value || typeof value !== "object")
        return undefined;
    const media = value;
    if (media.variant === "icon" && typeof media.name === "string") {
        return {
            variant: "icon",
            name: media.name,
            color: typeof media.color === "string" ? media.color : undefined,
        };
    }
    if (media.variant === "image" && typeof media.url === "string") {
        return {
            variant: "image",
            url: media.url,
            alt: typeof media.alt === "string" ? media.alt : undefined,
            round: typeof media.round === "boolean" ? media.round : undefined,
        };
    }
    return undefined;
}
export function SnapItem({ element: { props, children: childIds }, children, }) {
    const title = String(props.title ?? "");
    const description = props.description ? String(props.description) : undefined;
    const media = parseItemMedia(props.media);
    const colors = useSnapColors();
    const inBorderedGroup = useSnapItemGroupHasBorder();
    const inHorizontalStack = useSnapStackDirection() === "horizontal";
    const MediaIcon = media?.variant === "icon" ? ICON_MAP[media.name] : undefined;
    return (_jsxs(Item, { className: cn("gap-2 py-1.5", inBorderedGroup ? "px-2" : "px-0", 
        /** Horizontal: share width with peers. Vertical: don't fill column height. */
        inHorizontalStack && "flex-1"), style: {
            columnGap: 8,
            paddingInline: inBorderedGroup ? 8 : 0,
        }, children: [media?.variant === "icon" && MediaIcon && (_jsx(ItemMedia, { variant: "icon", className: "self-center translate-y-0", style: { alignSelf: "center", transform: "none" }, children: _jsx(MediaIcon, { size: 20, style: { color: colors.colorHex(media.color) } }) })), media?.variant === "image" && (_jsx(ItemMedia, { variant: "image", className: "self-center translate-y-0", style: {
                    alignSelf: "center",
                    borderRadius: media.round ? "9999px" : undefined,
                    transform: "none",
                }, children: _jsx("img", { src: media.url, alt: media.alt ?? "", className: "size-full object-cover" }) })), _jsxs(ItemContent, { className: "gap-0", children: [_jsx(ItemTitle, { style: { color: colors.text, fontSize: 14, lineHeight: "19px" }, children: title }), description && (_jsx(ItemDescription, { className: "mt-0 text-xs leading-snug", style: {
                            color: colors.textMuted,
                            fontSize: 12,
                            lineHeight: "16px",
                        }, children: description }))] }), childIds && childIds.length > 0 && (_jsx(ItemActions, { className: "gap-1.5 self-center", style: { alignSelf: "center", columnGap: 6 }, children: children }))] }));
}
