// @farcaster/snap@2.6.4/dist/react/components/text.js
"use client";
import { jsx as _jsx } from "react/jsx-runtime";
import { Text } from "@neynar/ui/typography";
import { cn } from "@neynar/ui/utils";
import { useSnapColors } from "../hooks/use-snap-colors.js";
import { useSnapStackDirection } from "../stack-direction-context.js";
const SIZE_MAP = {
    md: { textSize: "base" },
    sm: { textSize: "sm" },
};
export function SnapText({ element: { props }, }) {
    const content = String(props.content ?? "");
    const size = String(props.size ?? "md");
    const weight = props.weight
        ? String(props.weight)
        : undefined;
    const align = props.align ?? undefined;
    const config = SIZE_MAP[size] ?? SIZE_MAP.md;
    const colors = useSnapColors();
    const stackDir = useSnapStackDirection();
    const inHorizontalStack = stackDir === "horizontal";
    const maxLines = typeof props.maxLines === "number" ? props.maxLines : undefined;
    return (_jsx(Text, { size: config.textSize, weight: weight, align: align, className: cn(
        /**
         * Row peers hug content like RN `wrapRow` — `min-w-0 shrink` lets text wrap
         * inside a horizontal stack without forcing peers wide. In a vertical stack
         * the `<p>` already fills its parent's width via `display: block`; avoid
         * `flex-1` here because `flex-grow: 1` on a vertical-flex child fills the
         * column's height, distributing siblings when the row is taller than its
         * content (e.g. text next to a tall image).
         */
        inHorizontalStack ? "min-w-0 shrink" : "min-w-0"), style: {
            color: colors.text,
            fontSize: size === "md" ? 15 : undefined,
            lineHeight: size === "sm" ? 1.35 : 1.4,
            ...(maxLines
                ? {
                    display: "-webkit-box",
                    WebkitBoxOrient: "vertical",
                    WebkitLineClamp: maxLines,
                    overflow: "hidden",
                }
                : {}),
        }, children: content }));
}
