// @farcaster/snap@2.6.4/dist/react/snap-view-core.js
"use client";
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { snapJsonRenderCatalog } from "../ui/index.js";
import { SnapCatalogView } from "./catalog-renderer.js";
import { SnapPreviewAccentProvider } from "./accent-context.js";
import { SnapVersionProvider } from "./snap-version-context.js";
import { resolveSnapPaletteHex } from "./lib/resolve-palette-hex.js";
import { snapPreviewPrimaryCssProperties } from "./lib/preview-primary-css.js";
import { useCallback, useEffect, useMemo, useRef, useState, } from "react";
// ─── Internal helpers ──────────────────────────────────
export function applyStatePaths(model, changes) {
    if (!changes)
        return;
    const entries = Array.isArray(changes)
        ? changes.map((c) => [c.path, c.value])
        : Object.entries(changes);
    for (const [path, value] of entries) {
        const trimmed = path.startsWith("/") ? path : `/${path}`;
        const parts = trimmed.split("/").filter(Boolean);
        if (parts.length < 2)
            continue;
        const [top, ...rest] = parts;
        if (top === "inputs") {
            if (typeof model.inputs !== "object" || model.inputs === null) {
                model.inputs = {};
            }
            const inputs = model.inputs;
            if (rest.length === 1) {
                inputs[rest[0]] = value;
            }
            continue;
        }
        if (top === "theme") {
            if (typeof model.theme !== "object" || model.theme === null) {
                model.theme = {};
            }
            const theme = model.theme;
            if (rest.length === 1) {
                theme[rest[0]] = value;
            }
        }
    }
}
function withDefaultElementProps(spec) {
    if (!spec || typeof spec !== "object" || !("elements" in spec))
        return spec;
    const elements = spec.elements;
    if (!elements || typeof elements !== "object")
        return spec;
    let changed = false;
    const nextElements = {};
    for (const [id, element] of Object.entries(elements)) {
        if (element.props !== undefined) {
            nextElements[id] = element;
            continue;
        }
        changed = true;
        nextElements[id] = { ...element, props: {} };
    }
    return changed
        ? { ...spec, elements: nextElements }
        : spec;
}
const CONFETTI_COLORS = [
    "#907AA9",
    "#EC4899",
    "#3B82F6",
    "#10B981",
    "#F59E0B",
    "#EF4444",
    "#06B6D4",
];
const FIREWORK_COLORS = [
    "#FFD700",
    "#FF6B6B",
    "#4ECDC4",
    "#C4A7E7",
    "#F6C177",
    "#EBBCBA",
    "#9CCFD8",
    "#fff",
];
function ConfettiOverlay() {
    const pieces = useMemo(() => Array.from({ length: 80 }, (_, i) => {
        const driftX = (Math.random() - 0.5) * 120;
        return {
            id: i,
            left: Math.random() * 100,
            delay: Math.random() * 1.2,
            duration: 2.8 + Math.random() * 1.8,
            color: CONFETTI_COLORS[Math.floor(Math.random() * CONFETTI_COLORS.length)],
            size: 6 + Math.random() * 8,
            rotation: Math.random() * 360,
            isCircle: Math.random() > 0.6,
            driftX,
            driftMid: -driftX * 0.4,
        };
    }), []);
    return (_jsxs("div", { style: {
            position: "absolute",
            inset: 0,
            overflow: "hidden",
            pointerEvents: "none",
            zIndex: 20,
        }, children: [pieces.map(({ id, left, delay, duration, color, size, rotation, isCircle, driftX, driftMid }) => (_jsx("div", { style: {
                    position: "absolute",
                    left: `${left}%`,
                    top: -20,
                    width: size,
                    height: isCircle ? size : size * 0.5,
                    backgroundColor: color,
                    borderRadius: isCircle ? "50%" : 2,
                    transform: `rotateZ(${rotation}deg)`,
                    animation: `confettiFall ${duration}s cubic-bezier(0.25,0,0.75,1) ${delay}s forwards`,
                    "--dx": `${driftX}px`,
                    "--dm": `${driftMid}px`,
                } }, id))), _jsx("style", { children: `@keyframes confettiFall{
        0%  {top:-20px;opacity:1;transform:rotateZ(0deg) rotateY(0deg) translateX(0)}
        20% {transform:rotateZ(144deg) rotateY(60deg) translateX(var(--dm))}
        40% {transform:rotateZ(288deg) rotateY(120deg) translateX(0)}
        60% {opacity:1;transform:rotateZ(432deg) rotateY(200deg) translateX(calc(-1 * var(--dm)))}
        80% {transform:rotateZ(576deg) rotateY(280deg) translateX(var(--dx))}
        100%{top:110%;opacity:0;transform:rotateZ(720deg) rotateY(360deg) translateX(var(--dx))}
      }` })] }));
}
function FireworksOverlay() {
    const bursts = useMemo(() => Array.from({ length: 5 }, (_, b) => ({
        id: b,
        x: 15 + Math.random() * 70,
        y: 10 + Math.random() * 50,
        delay: b * 0.5 + Math.random() * 0.2,
        particles: Array.from({ length: 24 }, (_, p) => {
            const angle = (p / 24) * Math.PI * 2 + (Math.random() - 0.5) * 0.2;
            const dist = 55 + Math.random() * 60;
            return {
                id: p,
                vx: Math.cos(angle) * dist,
                vy: Math.sin(angle) * dist,
                color: FIREWORK_COLORS[Math.floor(Math.random() * FIREWORK_COLORS.length)],
                size: 3 + Math.random() * 3,
            };
        }),
    })), []);
    return (_jsxs("div", { style: {
            position: "absolute",
            inset: 0,
            overflow: "hidden",
            pointerEvents: "none",
            zIndex: 20,
        }, children: [bursts.map(({ id: bid, x, y, delay, particles }) => (_jsxs("div", { children: [_jsx("div", { style: {
                            position: "absolute",
                            left: `${x}%`,
                            top: `${y}%`,
                            width: 12,
                            height: 12,
                            borderRadius: "50%",
                            backgroundColor: "#fff",
                            transform: "translate(-50%,-50%)",
                            animation: `fwFlash 0.4s ease-out ${delay}s both`,
                            opacity: 0,
                        } }), particles.map(({ id: pid, vx, vy, color, size }) => (_jsx("div", { style: {
                            position: "absolute",
                            left: `${x}%`,
                            top: `${y}%`,
                            width: size,
                            height: size,
                            borderRadius: "50%",
                            backgroundColor: color,
                            transform: "translate(-50%,-50%)",
                            animation: `fwBurst 1s cubic-bezier(0.2,0,0.8,1) ${delay}s forwards`,
                            opacity: 0,
                            "--vx": `${vx}px`,
                            "--vy": `${vy}px`,
                        } }, pid)))] }, bid))), _jsx("style", { children: `
        @keyframes fwFlash{
          0%  {opacity:0;transform:translate(-50%,-50%) scale(0)}
          25% {opacity:1;transform:translate(-50%,-50%) scale(2.5)}
          100%{opacity:0;transform:translate(-50%,-50%) scale(5)}
        }
        @keyframes fwBurst{
          0%  {opacity:1;transform:translate(-50%,-50%) translate(0,0) scale(1)}
          65% {opacity:1}
          100%{opacity:0;transform:translate(-50%,-50%) translate(var(--vx),var(--vy)) scale(0)}
        }
      ` })] }));
}
export function SnapLoadingOverlay({ appearance, accentHex, active, }) {
    const isDark = appearance === "dark";
    const tint = isDark ? "rgba(0, 0, 0, 0.1)" : "rgba(255, 255, 255, 0.2)";
    const trackColor = isDark
        ? "rgba(255, 255, 255, 0.12)"
        : "rgba(15, 23, 42, 0.1)";
    return (_jsxs("div", { style: {
            position: "absolute",
            inset: 0,
            display: "flex",
            alignItems: "center",
            justifyContent: "center",
            zIndex: 10,
            background: tint,
            backdropFilter: active ? "blur(10px) saturate(1.05)" : "none",
            WebkitBackdropFilter: active
                ? "blur(10px) saturate(1.05)"
                : "none",
            opacity: active ? 1 : 0,
            pointerEvents: active ? "auto" : "none",
            transition: "opacity 0.28s ease, backdrop-filter 0.28s ease",
        }, "aria-hidden": !active, "aria-busy": active ? true : undefined, "aria-live": active ? "polite" : undefined, "aria-label": active ? "Loading" : undefined, children: [_jsx("div", { "data-snap-loading-spinner": true, style: {
                    width: 30,
                    height: 30,
                    borderRadius: "50%",
                    border: `2.5px solid ${trackColor}`,
                    borderTopColor: accentHex,
                    opacity: 0.88,
                    animation: "snapViewSpin 0.75s linear infinite",
                    flexShrink: 0,
                } }), _jsx("style", { children: `
        @keyframes snapViewSpin {
          to { transform: rotate(360deg); }
        }
        @media (prefers-reduced-motion: reduce) {
          [data-snap-loading-spinner] {
            animation: none;
            border-top-color: ${accentHex};
            opacity: 0.75;
          }
        }
      ` })] }));
}
const PALETTE = [
    "gray",
    "blue",
    "red",
    "amber",
    "green",
    "teal",
    "purple",
    "pink",
];
// ─── SnapViewCore ────────────────────────────────────
// Shared rendering logic used by both v1 and v2.
export function SnapViewCore({ snap, handlers, loading = false, appearance = "dark", loadingOverlay, }) {
    const spec = useMemo(() => withDefaultElementProps(snap.ui), [snap.ui]);
    const initialState = useMemo(() => spec.state ?? { inputs: {} }, [spec]);
    const stateRef = useRef(initialState);
    useEffect(() => {
        stateRef.current = {
            inputs: {
                ...(initialState.inputs ?? {}),
            },
            theme: {
                ...(initialState.theme ?? {}),
            },
        };
    }, [initialState]);
    useEffect(() => {
        const catalogResult = snapJsonRenderCatalog.validate(spec);
        if (!catalogResult.success) {
            // eslint-disable-next-line no-console
            console.warn("[Snap] catalog validation issues:", catalogResult.error);
        }
    }, [spec]);
    const [pageKey, setPageKey] = useState(0);
    useEffect(() => {
        setPageKey((k) => k + 1);
    }, [spec]);
    const showConfetti = snap.effects?.includes("confetti") ?? false;
    const showFireworks = snap.effects?.includes("fireworks") ?? false;
    const [confettiKey, setConfettiKey] = useState(0);
    const [fireworksKey, setFireworksKey] = useState(0);
    useEffect(() => {
        if (showConfetti)
            setConfettiKey((k) => k + 1);
        if (showFireworks)
            setFireworksKey((k) => k + 1);
    }, [showConfetti, showFireworks, snap]);
    const accentName = snap.theme?.accent ?? "purple";
    const accentHex = useMemo(() => resolveSnapPaletteHex(accentName, appearance), [accentName, appearance]);
    const previewSurfaceStyle = useMemo(() => {
        const vars = {};
        for (const c of PALETTE)
            vars[`--snap-color-${c}`] = resolveSnapPaletteHex(c, appearance);
        return {
            ...snapPreviewPrimaryCssProperties(accentName, appearance),
            ...vars,
        };
    }, [accentName, appearance]);
    const handleAction = useCallback((name, params) => {
        const inputs = (stateRef.current.inputs ?? {});
        const p = (params ?? {});
        switch (name) {
            case "submit":
                handlers.submit(String(p.target ?? ""), inputs);
                break;
            case "open_url":
                handlers.open_url(String(p.target ?? ""));
                break;
            case "open_snap":
                handlers.open_snap(String(p.target ?? ""));
                break;
            case "open_mini_app":
                handlers.open_mini_app(String(p.target ?? ""));
                break;
            case "view_cast":
                handlers.view_cast({ hash: String(p.hash ?? "") });
                break;
            case "view_profile":
                handlers.view_profile({ fid: Number(p.fid ?? 0) });
                break;
            case "compose_cast":
                handlers.compose_cast({
                    text: p.text ? String(p.text) : undefined,
                    channelKey: p.channelKey ? String(p.channelKey) : undefined,
                    embeds: Array.isArray(p.embeds)
                        ? p.embeds
                        : undefined,
                });
                break;
            case "view_token":
                handlers.view_token({ token: String(p.token ?? "") });
                break;
            case "send_token":
                handlers.send_token({
                    token: String(p.token ?? ""),
                    amount: p.amount ? String(p.amount) : undefined,
                    recipientFid: p.recipientFid ? Number(p.recipientFid) : undefined,
                    recipientAddress: p.recipientAddress
                        ? String(p.recipientAddress)
                        : undefined,
                });
                break;
            case "swap_token":
                handlers.swap_token({
                    sellToken: p.sellToken ? String(p.sellToken) : undefined,
                    buyToken: p.buyToken ? String(p.buyToken) : undefined,
                });
                break;
            default:
                break;
        }
    }, [handlers]);
    return (_jsxs("div", { style: { position: "relative", width: "100%" }, children: [showConfetti && _jsx(ConfettiOverlay, {}, confettiKey), showFireworks && _jsx(FireworksOverlay, {}, fireworksKey), loadingOverlay === undefined ? (_jsx(SnapLoadingOverlay, { appearance: appearance, accentHex: accentHex, active: loading })) : loading ? (_jsx(_Fragment, { children: loadingOverlay })) : null, _jsx("div", { style: previewSurfaceStyle, children: _jsx(SnapPreviewAccentProvider, { pageAccent: snap.theme?.accent, appearance: appearance, children: _jsx(SnapVersionProvider, { value: snap.version === "2.0" ? "2.0" : "1.0", children: _jsx(SnapCatalogView, { spec: spec, state: initialState, loading: false, onStateChange: (changes) => {
                                applyStatePaths(stateRef.current, changes);
                            }, onAction: handleAction }, pageKey) }) }) })] }));
}
