// @farcaster/snap@2.6.4/dist/react-native/v2/snap-view.js
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { useEffect, useMemo, useState } from "react";
import { Platform, Pressable, StyleSheet, Text, View } from "react-native";
import { SnapThemeProvider, useSnapTheme } from "../theme.js";
import { SnapLoadingOverlay, SnapViewCoreInner, resolveAccentHex, } from "../snap-view-core.js";
import { validateSnapResponse, } from "@farcaster/snap";
import { getSnapExpansionState, SNAP_MAX_HEIGHT } from "../expand-state.js";
// ─── Constants ───────────────────────────────────────
const SNAP_WARNING_HEIGHT = 700;
const SHOW_MORE_OVERHANG = 14;
// ─── Validation fallback ─────────────────────────────
function SnapValidationFallback({ message }) {
    const { colors } = useSnapTheme();
    return (_jsx(View, { style: fallbackStyles.container, children: _jsx(Text, { style: [fallbackStyles.text, { color: colors.textSecondary }], children: message ? `Unable to render snap: ${message}` : "Unable to render snap" }) }));
}
const fallbackStyles = StyleSheet.create({
    container: {
        width: "100%",
        padding: 16,
        alignItems: "center",
        justifyContent: "center",
    },
    text: {
        fontSize: 14,
    },
});
// ─── SnapViewV2 (with validation) ────────────────────
export function SnapViewV2Inner({ snap, handlers, loading = false, onValidationError, validationErrorFallback, loadingOverlay, }) {
    const validation = useMemo(() => validateSnapResponse(snap), [snap]);
    const valid = validation.valid;
    const validationMessage = validation.issues[0]?.message;
    useEffect(() => {
        if (!valid) {
            if (onValidationError) {
                onValidationError(validation);
            }
            else {
                // eslint-disable-next-line no-console
                console.warn("[Snap] validation issues:", validation.issues);
            }
        }
    }, [valid, validation, onValidationError]);
    if (!valid) {
        if (validationErrorFallback === null)
            return null;
        return (_jsx(_Fragment, { children: validationErrorFallback ?? _jsx(SnapValidationFallback, { message: validationMessage }) }));
    }
    return (_jsx(SnapViewCoreInner, { snap: snap, handlers: handlers, loading: loading, loadingOverlay: loadingOverlay }));
}
export function SnapViewV2({ snap, handlers, loading = false, appearance = "dark", colors, onValidationError, validationErrorFallback, loadingOverlay, }) {
    return (_jsx(SnapThemeProvider, { appearance: appearance, colors: colors, children: _jsx(SnapViewV2Inner, { snap: snap, handlers: handlers, loading: loading, onValidationError: onValidationError, validationErrorFallback: validationErrorFallback, loadingOverlay: loadingOverlay }) }));
}
// ─── SnapCardV2 (card frame + height limits) ─────────
function SnapCardV2Inner({ snap, handlers, loading, borderRadius, showOverflowWarning, onValidationError, validationErrorFallback, actionError, appearance, plain, loadingOverlay, forceExpanded, expandButtonLabel, onExpandPress, }) {
    const { colors, mode } = useSnapTheme();
    const accentHex = resolveAccentHex(snap.theme?.accent, mode);
    const [contentHeight, setContentHeight] = useState(0);
    const [isExpanded, setIsExpanded] = useState(false);
    useEffect(() => {
        setIsExpanded(false);
        setContentHeight(0);
    }, [snap]);
    const expansion = getSnapExpansionState({
        contentHeight,
        internalExpanded: isExpanded,
        forceExpanded,
        onExpandPress,
        expandButtonLabel,
        showOverflowWarning,
    });
    const expandButtonInsideCard = typeof onExpandPress === "function";
    const content = (_jsx(SnapViewV2Inner, { snap: snap, handlers: handlers, loading: loading, onValidationError: onValidationError, validationErrorFallback: validationErrorFallback, loadingOverlay: null }));
    if (plain) {
        return (_jsxs(_Fragment, { children: [_jsx(View, { style: expansion.clipped
                        ? { maxHeight: expansion.maxHeight, overflow: "hidden" }
                        : undefined, children: _jsx(View, { collapsable: false, onLayout: (e) => {
                            const nextHeight = Math.round(e.nativeEvent.layout.height);
                            setContentHeight((current) => expansion.clipped
                                ? Math.max(current, nextHeight)
                                : current === nextHeight
                                    ? current
                                    : nextHeight);
                        }, children: content }) }), loading
                    ? loadingOverlay === undefined
                        ? _jsx(SnapLoadingOverlay, { appearance: mode, accentHex: accentHex })
                        : loadingOverlay
                    : null, expansion.showButton ? (_jsx(View, { style: [cardStyles.expandRow, cardStyles.expandRowPlain], children: _jsx(Pressable, { style: ({ pressed }) => [
                            cardStyles.expandButton,
                            {
                                backgroundColor: pressed ? colors.mutedHover : colors.muted,
                            },
                        ], onPress: () => {
                            if (expansion.useInternalToggle) {
                                setIsExpanded((value) => !value);
                            }
                            else {
                                onExpandPress?.();
                            }
                        }, children: _jsx(Text, { style: [cardStyles.expandButtonText, { color: colors.text }], children: expansion.buttonLabel }) }) })) : null] }));
    }
    const overflowAmount = expansion.showOverflowWarning
        ? contentHeight - SNAP_MAX_HEIGHT
        : 0;
    const isDark = mode === "dark";
    const pillBg = isDark ? "rgba(40,40,40,0.92)" : "rgba(255,255,255,0.92)";
    const pillBgPressed = isDark ? "rgba(60,60,60,0.95)" : "rgba(240,240,240,0.95)";
    return (_jsxs(View, { style: {
            paddingBottom: expansion.showButton && !expandButtonInsideCard ? SHOW_MORE_OVERHANG : 0,
        }, children: [_jsxs(View, { style: { position: "relative" }, children: [_jsxs(View, { style: {
                            borderRadius,
                            borderWidth: 1,
                            borderColor: colors.border,
                            backgroundColor: colors.surface,
                            maxHeight: expansion.showOverflowWarning
                                ? undefined
                                : expansion.maxHeight,
                            overflow: "hidden",
                            minHeight: 120,
                        }, children: [_jsx(View, { collapsable: false, onLayout: (e) => {
                                    const nextHeight = Math.round(e.nativeEvent.layout.height);
                                    setContentHeight((current) => expansion.clipped
                                        ? Math.max(current, nextHeight)
                                        : current === nextHeight
                                            ? current
                                            : nextHeight);
                                }, style: cardStyles.body, children: content }), expansion.showOverflowWarning && contentHeight > SNAP_MAX_HEIGHT && (_jsxs(View, { style: { position: "absolute", top: SNAP_MAX_HEIGHT, left: 0, right: 0, height: overflowAmount, zIndex: 10, pointerEvents: "none" }, children: [_jsx(View, { style: { height: 1, borderTopWidth: 1, borderStyle: "dashed", borderColor: "rgba(255,100,100,0.6)" } }), _jsx(View, { style: { position: "absolute", top: -10, right: 4, backgroundColor: "rgba(0,0,0,0.7)", paddingHorizontal: 4, paddingVertical: 1, borderRadius: 3 }, children: _jsxs(Text, { style: { fontSize: 10, color: "rgba(255,100,100,0.7)", fontFamily: Platform.select({ ios: "Menlo", default: "monospace" }) }, children: [SNAP_MAX_HEIGHT, "px"] }) }), _jsx(View, { style: { flex: 1, backgroundColor: "rgba(255,50,50,0.15)" } })] })), loading
                                ? loadingOverlay === undefined
                                    ? _jsx(SnapLoadingOverlay, { appearance: mode, accentHex: accentHex })
                                    : loadingOverlay
                                : null] }), expansion.showButton ? (_jsx(View, { pointerEvents: "box-none", style: expandButtonInsideCard
                            ? cardStyles.expandFloatInset
                            : cardStyles.expandFloat, children: _jsx(Pressable, { style: ({ pressed }) => [
                                cardStyles.expandButton,
                                {
                                    backgroundColor: pressed ? pillBgPressed : pillBg,
                                    borderColor: colors.border,
                                },
                            ], onPress: () => {
                                if (expansion.useInternalToggle) {
                                    setIsExpanded((value) => !value);
                                }
                                else {
                                    onExpandPress?.();
                                }
                            }, children: _jsx(Text, { style: [cardStyles.expandButtonText, { color: colors.text }], children: expansion.buttonLabel }) }) })) : null] }), actionError && (_jsx(Text, { style: {
                    paddingHorizontal: 12,
                    paddingVertical: 8,
                    fontSize: 13,
                    color: appearance === "dark"
                        ? "rgba(255,100,100,0.9)"
                        : "rgba(200,0,0,0.8)",
                }, children: actionError }))] }));
}
export function SnapCardV2({ snap, handlers, loading = false, appearance = "dark", colors, borderRadius = 16, showOverflowWarning = false, onValidationError, validationErrorFallback, actionError, plain = false, loadingOverlay, forceExpanded, expandButtonLabel, onExpandPress, }) {
    return (_jsx(SnapThemeProvider, { appearance: appearance, colors: colors, children: _jsx(SnapCardV2Inner, { snap: snap, handlers: handlers, loading: loading, borderRadius: borderRadius, showOverflowWarning: showOverflowWarning, onValidationError: onValidationError, validationErrorFallback: validationErrorFallback, actionError: actionError, appearance: appearance, plain: plain, loadingOverlay: loadingOverlay, forceExpanded: forceExpanded, expandButtonLabel: expandButtonLabel, onExpandPress: onExpandPress }) }));
}
const cardStyles = StyleSheet.create({
    frameRing: { alignSelf: "stretch" },
    card: { borderWidth: 1, minHeight: 120, overflow: "hidden" },
    body: { paddingHorizontal: 8, paddingVertical: 8 },
    actionError: { paddingHorizontal: 12, paddingVertical: 8, fontSize: 13 },
    expandFloat: {
        position: "absolute",
        left: 0,
        right: 0,
        bottom: -14,
        height: 28,
        alignItems: "center",
        justifyContent: "center",
    },
    expandFloatInset: {
        position: "absolute",
        left: 0,
        right: 0,
        bottom: 10,
        height: 28,
        alignItems: "center",
        justifyContent: "center",
    },
    expandRow: {
        alignItems: "center",
    },
    expandRowPlain: {
        paddingTop: 8,
        alignItems: "center",
    },
    expandButton: {
        minWidth: 92,
        alignItems: "center",
        justifyContent: "center",
        borderRadius: 9999,
        borderWidth: 1,
        paddingHorizontal: 10,
        paddingVertical: 4,
    },
    expandButtonText: {
        fontSize: 12,
        lineHeight: 16,
        fontWeight: "600",
    },
    warningOverlay: {
        position: "absolute",
        top: SNAP_MAX_HEIGHT,
        left: 0,
        right: 0,
        bottom: 0,
        zIndex: 10,
    },
    warningLine: {
        height: 1,
        borderTopWidth: 1,
        borderStyle: "dashed",
        borderColor: "rgba(255,100,100,0.6)",
    },
    warningLabel: {
        position: "absolute",
        top: -10,
        right: 4,
        backgroundColor: "rgba(0,0,0,0.7)",
        paddingHorizontal: 4,
        paddingVertical: 1,
        borderRadius: 3,
    },
    warningLabelText: {
        fontSize: 10,
        color: "rgba(255,100,100,0.7)",
        fontFamily: Platform.select({ ios: "Menlo", default: "monospace" }),
    },
});
